openassessment/xblock/openassessmentblock.py

```diff
@@ -1106,11 +1106,23 @@ def is_closed(self, step=None, course_staff=None):
             datetime.datetime(2015, 3, 27, 22, 7, 38, 788861)
 
         """
-        submission_range = (self.submission_start, self.submission_due)
-        assessment_ranges = [
-            (asmnt.get('start'), asmnt.get('due'))
-            for asmnt in self.valid_assessments
-        ]
+        if self.due is not None \
+           and self.submission_due is not None \
+           and parse_date_value(self.due, self._) > parse_date_value(self.submission_due, self._):
+            submission_range = (self.submission_start, self.due)
+            assessment_ranges = []
+            for asmnt in self.valid_assessments:
+                asmnt_due = asmnt.get('due')
+                if asmnt_due is None or parse_date_value(self.due, self._) < parse_date_value(asmnt_due, self._):
+                    assessment_ranges.append((asmnt.get('start'), asmnt_due))
+                else:
+                    assessment_ranges.append((asmnt.get('start'), self.due))                
+        else:
+            submission_range = (self.submission_start, self.submission_due)
+            assessment_ranges = [
+                (asmnt.get('start'), asmnt.get('due'))
+                for asmnt in self.valid_assessments
+            ]
```

Comment:
This block is hard to read because of how nested it is, can we try refactoring it a bit?